src/peergos/server/tests/simulation/Simulator.java

```diff
@@ -212,61 +344,102 @@ public void run() {
+                    .collect(Collectors.toSet());
 
-        for (FileSystem fs : Arrays.asList(testFileSystem, referenceFileSystem)) {
 
-            Set<Path> paths = new HashSet<>();
-            fs.walk(paths::add);
+            FileSystem testFileSystem = fileSystems.getTestFileSystem(user);
+            FileSystem referenceFileSystem = fileSystems.getReferenceFileSystem(user);
 
-            Set<Path> expectedFilesAndFolders  =  new HashSet<>(expectedFiles);
-            expectedFilesAndFolders.addAll(simulatedDirectoryToFiles.keySet());
+            boolean isUserVerified = true;
+            for (FileSystem fs : Arrays.asList(testFileSystem, referenceFileSystem)) {
 
-            // extras?
-            Set<Path> extras = new HashSet<>(paths);
-            extras.removeAll(expectedFilesAndFolders);
+                Set<Path> paths = new HashSet<>();
+                fs.walk(paths::add);
 
-            for (Path extra : extras) {
-                LOG.info("filesystem " + fs + " has an extra path " + extra);
-                isVerified = false;
-            }
+                Set<Path> expectedFilesAndFolders = new HashSet<>(expectedFilesForUser);
+                expectedFilesAndFolders.addAll(dirToFiles.keySet());
 
-            // missing?
-            expectedFilesAndFolders.removeAll(paths);
-            for (Path missing : expectedFilesAndFolders) {
-                LOG.info("filesystem " + fs + " is missing  the path " + missing);
-                isVerified = false;
-            }
-        }
+                // extras?
+                Set<Path> extras = new HashSet<>(paths);
+                extras.removeAll(expectedFilesAndFolders);
 
-        // contents
-        for (Path path : expectedFiles) {
-            try {
-                byte[] testData = testFileSystem.read(path);
-                byte[] refData = referenceFileSystem.read(path);
-                if (!Arrays.equals(testData, refData)) {
-                    LOG.info("Path " + path + " has different contents between the file-systems");
-                    isVerified = false;
+                for (Path extra : extras) {
+                    LOG.info("filesystem " + fs + " has an extra path " + extra);
+                    isUserVerified = false;
                 }
-            } catch (Exception ex) {
-                LOG.info("Failed to read path + " + path);
-                isVerified = false;
+
+                // missing?
+                expectedFilesAndFolders.removeAll(paths);
+                for (Path missing : expectedFilesAndFolders) {
+                    LOG.info("filesystem " + fs + " is missing  the path " + missing);
+                    isUserVerified = false;
+                }
+            }
+
+            // contents
+            for (Path path : expectedFilesForUser) {
+                boolean verifyContents = verifyContents(user, path);
+                boolean sharingPermissionsAreVerified = verifySharingPermissions(user, path);
```

Comment:
do we also verify permissions for directories?

Reply:
not  yet - that is still `TODO`